// @carbon/web-components/es/components-react/tearsheet/tearsheet.js
/**
 * @license
 *
 * Copyright IBM Corp. 2019, 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
export {
  TEARSHEET_INFLUENCER_PLACEMENT,
  TEARSHEET_INFLUENCER_WIDTH,
  TEARSHEET_WIDTH,
} from '././defs.js';
export { default as CustomElement } from '../../components/tearsheet/tearsheet.js';
import createReactCustomElementType from '../../globals/wrappers/createReactCustomElementType.js';
import PropTypes from 'prop-types';
import { prefix } from '../../globals/settings.js';
export var descriptor = {
  onBeforeClose: {
    event: ''.concat(prefix, '-tearsheet-beingclosed'),
  },
  onClose: {
    event: ''.concat(prefix, '-tearsheet-closed'),
  },
  onNavigateBack: {
    event: ''.concat(prefix, '-tearsheet-header-navigate-back'),
  },
};
export var propTypes = {
  onBeforeClose: PropTypes.func,
  onClose: PropTypes.func,
  onNavigateBack: PropTypes.func,
};
const Component = createReactCustomElementType(
  ''.concat(prefix, '-tearsheet'),
  descriptor
);
Component.propTypes = propTypes;
export default Component;
