// @carbon/web-components/lib/components/tearsheet/defs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TEARSHEET_WIDTH = exports.TEARSHEET_INFLUENCER_WIDTH = exports.TEARSHEET_INFLUENCER_PLACEMENT = void 0;
/**
 * @license
 *
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
let TEARSHEET_INFLUENCER_PLACEMENT = exports.TEARSHEET_INFLUENCER_PLACEMENT = void 0;
(function (TEARSHEET_INFLUENCER_PLACEMENT) {
  TEARSHEET_INFLUENCER_PLACEMENT["RIGHT"] = "right";
  TEARSHEET_INFLUENCER_PLACEMENT["LEFT"] = "left";
})(TEARSHEET_INFLUENCER_PLACEMENT || (exports.TEARSHEET_INFLUENCER_PLACEMENT = TEARSHEET_INFLUENCER_PLACEMENT = {}));
let TEARSHEET_INFLUENCER_WIDTH = exports.TEARSHEET_INFLUENCER_WIDTH = void 0;
(function (TEARSHEET_INFLUENCER_WIDTH) {
  TEARSHEET_INFLUENCER_WIDTH["NARROW"] = "narrow";
  TEARSHEET_INFLUENCER_WIDTH["WIDE"] = "wide";
})(TEARSHEET_INFLUENCER_WIDTH || (exports.TEARSHEET_INFLUENCER_WIDTH = TEARSHEET_INFLUENCER_WIDTH = {}));
let TEARSHEET_WIDTH = exports.TEARSHEET_WIDTH = void 0;
(function (TEARSHEET_WIDTH) {
  TEARSHEET_WIDTH["NARROW"] = "narrow";
  TEARSHEET_WIDTH["WIDE"] = "wide";
})(TEARSHEET_WIDTH || (exports.TEARSHEET_WIDTH = TEARSHEET_WIDTH = {}));
