// @carbon/web-components/es/components/popover/defs.js
/**
 * @license
 *
 * Copyright IBM Corp. 2020, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * The alignment choices of popover.
 */
export let POPOVER_ALIGNMENT;
(function (POPOVER_ALIGNMENT) {
  POPOVER_ALIGNMENT["TOP"] = "top";
  POPOVER_ALIGNMENT["TOP_LEFT"] = "top-left";
  POPOVER_ALIGNMENT["TOP_RIGHT"] = "top-right";
  POPOVER_ALIGNMENT["TOP_START"] = "top-start";
  POPOVER_ALIGNMENT["TOP_END"] = "top-end";
  POPOVER_ALIGNMENT["BOTTOM"] = "bottom";
  POPOVER_ALIGNMENT["BOTTOM_LEFT"] = "bottom-left";
  POPOVER_ALIGNMENT["BOTTOM_RIGHT"] = "bottom-right";
  POPOVER_ALIGNMENT["BOTTOM_START"] = "bottom-start";
  POPOVER_ALIGNMENT["BOTTOM_END"] = "bottom-end";
  POPOVER_ALIGNMENT["LEFT"] = "left";
  POPOVER_ALIGNMENT["LEFT_BOTTOM"] = "left-bottom";
  POPOVER_ALIGNMENT["LEFT_TOP"] = "left-top";
  POPOVER_ALIGNMENT["LEFT_START"] = "left-start";
  POPOVER_ALIGNMENT["LEFT_END"] = "left-end";
  POPOVER_ALIGNMENT["RIGHT"] = "right";
  POPOVER_ALIGNMENT["RIGHT_BOTTOM"] = "right-bottom";
  POPOVER_ALIGNMENT["RIGHT_TOP"] = "right-top";
  POPOVER_ALIGNMENT["RIGHT_START"] = "right-start";
  POPOVER_ALIGNMENT["RIGHT_END"] = "right-end";
})(POPOVER_ALIGNMENT || (POPOVER_ALIGNMENT = {}));
