// @carbon/web-components/es/components/notification/actionable-notification-button.js
import _decorate from "@babel/runtime/helpers/esm/decorate";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _get from "@babel/runtime/helpers/esm/get";
function _superPropGet(t, e, r, o) { var p = _get(_getPrototypeOf(1 & o ? t.prototype : t), e, r); return 2 & o ? function (t) { return p.apply(r, t); } : p; }
/**
 * @license
 *
 * Copyright IBM Corp. 2019, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from '../../globals/settings';
import CDSButton from '../button/button';
import styles from "././actionable-notification.css.js";
import { carbonElement as customElement } from '../../globals/decorators/carbon-element';

/**
 * Actionable notification action button.
 *
 * @element cds-actionable-notification-button
 */
let CDSActionableNotificationButton = _decorate([customElement(`${prefix}-actionable-notification-button`)], function (_initialize, _CDSButton) {
  class CDSActionableNotificationButton extends _CDSButton {
    constructor(...args) {
      super(...args);
      _initialize(this);
    }
  }
  return {
    F: CDSActionableNotificationButton,
    d: [{
      kind: "method",
      key: "update",
      value: function update(changedProperties) {
        var _getElementById;
        _superPropGet(CDSActionableNotificationButton, "update", this, 3)([changedProperties]);
        (_getElementById = this.shadowRoot.getElementById('button')) === null || _getElementById === void 0 || _getElementById.classList.add(`${prefix}--actionable-notification__action-button`);
        this.setAttribute('size', 'sm');
      }
    }, {
      kind: "field",
      static: true,
      key: "styles",
      value() {
        return styles;
      }
    }]
  };
}, CDSButton);
export default CDSActionableNotificationButton;
