// @carbon/web-components/es/components/ui-shell/header-side-nav-items.js
import _decorate from "@babel/runtime/helpers/esm/decorate";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _get from "@babel/runtime/helpers/esm/get";
let _ = t => t,
  _t;
function _superPropGet(t, e, r, o) { var p = _get(_getPrototypeOf(1 & o ? t.prototype : t), e, r); return 2 & o ? function (t) { return p.apply(r, t); } : p; }
/**
 * @license
 *
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { LitElement, html } from 'lit';
import { property } from 'lit/decorators.js';
import { carbonElement as customElement } from '../../globals/decorators/carbon-element';
import { prefix } from '../../globals/settings';
import styles from "././side-nav.css.js";

/**
 * Header Side Nav Items section
 *
 * @element cds-header-side-nav-items
 */ /**
     * Header Side Nav Items section
     *
     * @element cds-header-side-nav-items
     */
let CDSHeaderSideNavItems = _decorate([customElement(`${prefix}-header-side-nav-items`)], function (_initialize, _LitElement) {
  class CDSHeaderSideNavItems extends _LitElement {
    constructor(...args) {
      super(...args);
      _initialize(this);
    }
  }
  return {
    F: CDSHeaderSideNavItems,
    d: [{
      kind: "field",
      decorators: [property({
        type: Boolean,
        attribute: 'has-divider'
      })],
      key: "hasDivider",
      value() {
        return false;
      }
    }, {
      kind: "method",
      key: "connectedCallback",
      value:
      /**
       * Optionally specify if container will have a bottom divider to differentiate
       * between original sidenav items and header menu items. False by default.
       */

      function connectedCallback() {
        if (!this.hasAttribute('role')) {
          this.setAttribute('role', 'list');
        }
        _superPropGet(CDSHeaderSideNavItems, "connectedCallback", this, 3)([]);
      }
    }, {
      kind: "method",
      key: "render",
      value: function render() {
        return html(_t || (_t = _`<slot></slot>`));
      }
    }, {
      kind: "field",
      static: true,
      key: "styles",
      value() {
        return styles;
      }
    }]
  };
}, LitElement);
export default CDSHeaderSideNavItems;
