// @carbon/web-components/es/components/ui-shell/switcher-divider.js
import _decorate from "@babel/runtime/helpers/esm/decorate";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _get from "@babel/runtime/helpers/esm/get";
function _superPropGet(t, e, r, o) { var p = _get(_getPrototypeOf(1 & o ? t.prototype : t), e, r); return 2 & o ? function (t) { return p.apply(r, t); } : p; }
/**
 * @license
 *
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { LitElement } from 'lit';
import { prefix } from '../../globals/settings';
import styles from "././header.css.js";
import { carbonElement as customElement } from '../../globals/decorators/carbon-element';

/**
 * A divider in switcher.
 *
 * @element cds-switcher-divider
 */
let CDSSwitcherDivider = _decorate([customElement(`${prefix}-switcher-divider`)], function (_initialize, _LitElement) {
  class CDSSwitcherDivider extends _LitElement {
    constructor(...args) {
      super(...args);
      _initialize(this);
    }
  }
  return {
    F: CDSSwitcherDivider,
    d: [{
      kind: "method",
      key: "connectedCallback",
      value: function connectedCallback() {
        if (!this.hasAttribute('role')) {
          this.setAttribute('role', 'separator');
        }
        _superPropGet(CDSSwitcherDivider, "connectedCallback", this, 3)([]);
      }
    }, {
      kind: "field",
      static: true,
      key: "styles",
      value() {
        return styles;
      }
    }]
  };
}, LitElement);
export default CDSSwitcherDivider;
