// @carbon/web-components/es/components/ai-skeleton/ai-skeleton-placeholder.js
import _decorate from "@babel/runtime/helpers/esm/decorate";
let _ = t => t,
  _t;
/**
 * @license
 *
 * Copyright IBM Corp. 2019, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { LitElement, html } from 'lit';
import { prefix } from '../../globals/settings';
import styles from "././ai-skeleton.css.js";
import '../skeleton-placeholder/skeleton-placeholder';
import { carbonElement as customElement } from '../../globals/decorators/carbon-element';

/**
 * AI skeleton placeholder.
 *
 * @element cds-ai-skeleton-placeholder
 */
let CDSAISkeletonPlaceholder = _decorate([customElement(`${prefix}-ai-skeleton-placeholder`)], function (_initialize, _LitElement) {
  class CDSAISkeletonPlaceholder extends _LitElement {
    constructor(...args) {
      super(...args);
      _initialize(this);
    }
  }
  return {
    F: CDSAISkeletonPlaceholder,
    d: [{
      kind: "method",
      key: "render",
      value: function render() {
        return html(_t || (_t = _`<cds-skeleton-placeholder
      optional-classes="${0}--skeleton__placeholder--ai"></cds-skeleton-placeholder>`), prefix);
      }
    }, {
      kind: "field",
      static: true,
      key: "styles",
      value() {
        return styles;
      }
    }]
  };
}, LitElement);
export default CDSAISkeletonPlaceholder;
